packages/jest-mock/src/index.ts

```diff
@@ -768,6 +780,34 @@ export class ModuleMocker {
         return f;
       };
 
+      f.withImplementation = <R extends WithImplementationCallbackReturn>(
+        fn: UnknownFunction,
+        callback: () => R,
+        // @ts-expect-error: Type guards are not advanced enough for this use case
+      ): R extends WithImplementationAsyncCallbackReturn
+        ? Promise<void>
+        : undefined => {
```

Comment:
Oh.. Forgot to add this part as well (just move `function withImplementation` to some better place): 

```suggestion
      f.withImplementation = withImplementation.bind(this);

      function withImplementation(
        fn: UnknownFunction,
        callback: () => void,
      ): void;
      function withImplementation(
        fn: UnknownFunction,
        callback: () => Promise<unknown>,
      ): Promise<void>;
      function withImplementation(
        this: ModuleMocker,
        fn: UnknownFunction,
        callback: (() => void) | (() => Promise<unknown>),
      ): void | Promise<void> {
```